rs/registry/canister/src/mutations/routing_table.rs

```diff
@@ -31,18 +40,98 @@ impl std::fmt::Display for GetSubnetForCanisterError {
+        .peekable();
+    let mut new_it = new_rt
+        .iter()
+        .map(|(range, &subnet)| ((range.start, subnet), range.end))
+        .peekable();
+
+    let mut mutations = vec![];
+
+    let range_key = |key: (CanisterId, SubnetId)| -> Vec<u8> {
+        make_canister_range_key(key.0, key.1).as_bytes().to_vec()
+    };
+
+    let create_entry = |key: (CanisterId, SubnetId), range_end: CanisterId| -> Vec<u8> {
+        pb::routing_table::Entry {
+            range: Some(pb::CanisterIdRange {
+                start_canister_id: Some(key.0.into()),
+                end_canister_id: Some(range_end.into()),
+            }),
+            subnet_id: Some(pb_subnet_id(key.1)),
+        }
+        .encode_to_vec()
+    };
+
+    while old_it.peek().is_some() || new_it.peek().is_some() {
+        match (old_it.peek(), new_it.peek()) {
+            (Some(&(o_key, o_end)), Some(&(n_key, n_end))) => match o_key.cmp(&n_key) {
+                Ordering::Less => {
+                    mutations.push(delete(range_key(o_key)));
+                    old_it.next();
+                }
+                Ordering::Greater => {
+                    mutations.push(upsert(range_key(n_key), create_entry(n_key, n_end)));
+                    new_it.next();
+                }
+                Ordering::Equal => {
+                    // only change differences, these mutations take space.
+                    if o_end != n_end {
+                        mutations.push(upsert(range_key(n_key), create_entry(n_key, n_end)));
+                    }
+                    old_it.next();
+                    new_it.next();
+                }
+            },
+            (Some(&(o_key, _)), None) => {
+                mutations.push(delete(range_key(o_key)));
+                old_it.next();
+            }
+            (None, Some(&(n_key, n_end))) => {
+                mutations.push(upsert(range_key(n_key), create_entry(n_key, n_end)));
+                new_it.next();
+            }
+            (None, None) => unreachable!(),
+        }
+    }
+
+    mutations
+}
+
+fn pb_subnet_id(subnet_id: SubnetId) -> ic_protobuf::types::v1::SubnetId {
+    ic_protobuf::types::v1::SubnetId {
+        principal_id: Some(ic_protobuf::types::v1::PrincipalId {
+            raw: subnet_id.get().as_slice().to_vec(),
+        }),
+    }
+}
+
 pub(crate) fn routing_table_into_registry_mutation(
+    registry: &Registry,
     routing_table: RoutingTable,
-    mutation_type: i32,
-) -> RegistryMutation {
-    let routing_table = pb::RoutingTable::from(routing_table);
-    let mut buf = vec![];
-    routing_table.encode(&mut buf).unwrap();
-    RegistryMutation {
-        mutation_type,
-        key: make_routing_table_record_key().as_bytes().to_vec(),
-        value: buf,
+) -> Vec<RegistryMutation> {
+    let mut mutations = vec![];
+
+    if is_canister_ranges_routing_map_storage_enabled() {
+        let old = registry.get_routing_table_or_panic(registry.latest_version());
```

Comment:
This needs to get the table from get_routing_table_from_canister_range_records_or_panic